### ISSUE TITLE
Fails to type check argument to subscript operation

### ISSUE THREAD

**Issue by JelleZijlstra** [microsoft/pyright#8183]
Code sample:

```python
from typing import SupportsIndex, Protocol

class MySupportsIndex(Protocol):
    def __index__(self) -> int: ...

class Foo:
    def __getitem__(self, k: int) -> str: return "x"

def f(foo: Foo, key: SupportsIndex, key2: MySupportsIndex, key3: object):
    foo[key]
    foo[key2]
    foo.__getitem__(key)
    foo.__getitem__(key2)
```

I would expect an error for every line in `f()`, but pyright only gives an error on the last two lines (https://pyright-play.net/?strict=true&code=MQAgLgpgzmD6DGBDKEB0AHAngKAGYCcB7AW3E3QEsA7AcxAuPUPzBAGUBXdJlqASSoATCAA8ANCAAKRMIXiEANtmzwFyKCACymTt2Zh%2BQ0QApphWfIUBKAFzYQDkMNwhYsasJFvjKBbisgALQAfPRUYDYgqNHKquogAGKEhHaOThAubjQQYBSQxN6%2BuBIA1pHUYAEhIDD4kfg5HPhUIABEIq3KziC4xrjJkUmEpRCYkbo8BgKeI5gATJHaE-qGMyAlowDMkYQARgBWEPCVqY79hADaG5gAuvZnyVejc3dp56hZOXkQBbDG11Z7g53p9cvlvNc5oCgA). Mypy errors on all lines as I'd expect: https://mypy-play.net/?mypy=latest&python=3.12&gist=42a545701637e0cdbce6cfb9b4fa3d77.

**Comment by erictraut** [comment]
Yeah, pyright currently checks for the presence of a `__getitem__`, but it doesn't validate that the call can be made without a type violation. It should do this.

**Comment by erictraut** [comment]
Maybe my understanding of indexing isn't correct. I thought that the runtime first tried to invoke `__getitem__`. If that fails, it then checks whether the subscript object supports an `__index__` call. If so, it calls that to retrieve an `int` value, and then it attempts to call the `__getitem__` again with this `int` value. Is my understanding incorrect? I'm struggling to find good documentation for this behavior.

**Comment by JelleZijlstra** [comment]
For certain classes implemented in C, Python does automatically go through `__index__`. However, such classes should be annotated explicitly (`def __getitem__(self, index: SupportsIndex, /)`. For classes implemented in Python, this does not happen.

**Comment by erictraut** [comment]
Ah, makes sense. It's nice when a bug fix simply involves deleting a bunch of special-case code. :)

**Comment by erictraut** [comment]
This is included in pyright 1.1.369.
